### PR TITLE
Adds serialization

### LINKED ISSUE #51: Add point compression + canonical serialization for `AffineCurve` elements
This is one of the critical missing pieces for this library. A sketch of the design I have in mind is to make `CanonicalSerialize` and `CanonicalDeserialize` traits:
```rust
pub trait CanonicalSerialize {
	type Output: Iterator<Item = u8>
	fn serialize(&self) -> Output;
}

pub trait CanonicalDeserialize: CanonicalSerialize {
	fn deserialize(&Self::Output) -> Self;
}
```
The idea would then be to implement these directly for the final instantiated curves, e.g. for `Bls12_381_G1`. (I don't see an easy way to have a generic implementation that works for all base fields.)

### PR DESCRIPTION
**Pull request by kobigurk**
Closes #51.

### CODE REVIEW

File: algebra/src/serialize/error.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#[derive(Debug)]
+pub enum SerializationError {
+    /// During serialization, the output buffer was of the wrong size.
+    BufferWrongSize,
+    /// During serialization, we didn't have enough space to write extra info.
+    NotEnoughSpace,
+    /// During serialization, the data was invalid.
+    InvalidData,
+    /// During serialization, we countered an I/O error.
+    IoError(io::Error),
```
Comment (Pratyush) [inline algebra/src/serialize/error.rs:17]: We don't need this variant I think?

File: algebra/src/serialize/mod.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+        impl<P: $params> CanonicalDeserialize for GroupAffine<P> {
+            fn deserialize(bytes: &[u8], _: &mut [bool]) -> Result<Self, crate::serialize::SerializationError>
+                where Self: Sized {
+                let mut extra_info_buf = [false; 2];
+                let x = P::BaseField::deserialize(bytes, &mut extra_info_buf)?;
+                if extra_info_buf[1] {
+                    return Ok(Self::zero())
+                }
+                GroupAffine::<P>::get_point_from_x(x, extra_info_buf[0])
```
Comment (Pratyush) [inline algebra/src/serialize/mod.rs]: Should we add checks for ensuring the point is in the prime subgroup here?

File: algebra/src/serialize/mod.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        impl<P: $params> CanonicalDeserialize for GroupAffine<P> {
+            fn deserialize(bytes: &[u8], _: &mut [bool]) -> Result<Self, crate::serialize::SerializationError>
+                where Self: Sized {
+                let mut extra_info_buf = [false; 1];
+                let x = P::BaseField::deserialize(bytes, &mut extra_info_buf)?;
+                if x == P::BaseField::zero() {
+                    return Ok(Self::zero())
+                }
+                GroupAffine::<P>::get_point_from_x(x, extra_info_buf[0])
```
Comment (Pratyush) [inline algebra/src/serialize/mod.rs]: Should we add a prime-subgroup check here?

**Review by Pratyush — APPROVED** [review]

File: algebra/src/serialize/error.rs (hunk 1 above)
Comment (kobigurk) [reply algebra/src/serialize/error.rs:17]: The read/write use this variant

**Comment by Pratyush** [comment]
LGTM, I'll let CI run

**Pull request merged**